keps/sig-windows/2258-kubectl-service-log-viewer/README.md

```diff
@@ -0,0 +1,213 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+A Kubernetes cluster administrator has to log in to the relavant control-plane
+or worker nodes to view the logs of the API server, kubelet etc. Or they would
+have to implement a client side reader. A simpler and more elegant method would
+be to allow them to use the kubectl CLI to also view these logs similar to
+using it for other interactions with the cluster.
+
+## Motivation
+
+Troubleshooting issues with control-plane and worker nodes typically requires 
+a cluster administrator to SSH into the nodes for debugging. While certain
+issues will require being on the node, issues with the kube-proxy or kubelet,
+to name a couple, could be solved by perusing their logs. However this
+too requires the administrator to SSH access into the nodes. Giving a way for
+them to view the logs using kubectl will significantly simplify their
+troubleshooting.
+
+
+### Goals
+Provide a cluster administrator with a streaming view of logs using kubectl
+without them having to implement a client side reader or logging into the node.
+This would work for:
+- Linux worker and control plane nodes using a distribution that have systemd
+  / journald support
+- Windows worker nodes (all supported variants)
+
+### Non-Goals
+Providing support for non-systemd Linux distributions.
+
+## Proposal
+
+### Implement client for logs endpoint viewer (OS agnostic)
+Implement a client for the `/var/log/` kubelet endpoint viewer.
+
+```
+# See what logs are available in masters in /var/log/
+kubectl node-logs --role master --path=/
+
+# Display cron log file (/var/log/cron) from all masters
+kubectl node-logs --role master --path=cron
+```
+
+### Linux distros with systemd / journald
+Suplement the the `/var/log/` endpoint viewer on the kubelet with a thin shim over
+the `journal` directory that shells out to journalctl. Then add a new option to 
+kubectl called `node-logs` that will retrieve it.
+
+```
+# Show kubelet and crio journal logs from all masters
+kubectl node-logs node-logs --role master --path journal -u kubelet -u crio
+```
+
+### Linux distros with systemd / journald
+Running the new "kubectl node-logs" command against nodes that do not use
+systemd / journald should return "OS not supported".
+
+### Windows
+Reuse the kubelet API for querying the Linux journal for invoking the `Get-WinEvent`
```

Comment:
Windows kubelets may log to just a regular file. Would there be some logic here to figure out whether the log type configured and stream them out based on the configuration?

Reply:
I don't think we've planned for any detection logic for this case.
One thing to note is that if the kubelet (or any other process) logs to `C:\var\log\somepath` on Windows, those log files can be streamed by the kubelet's existing log server, and one part of this proposal is implementing a client command for it.

Reply:
@ddebroy I have sort of addressed this in my next commit. PTAL.